# マチンの公式の見つけ方

マチンの公式のような公式を見つける、おそらく最も簡単な方法を紹介します。まず適当に自然数 $k \in \mathbb{N}$ を選びます。次に

$$m \arctan \frac{1}{k} \leq \frac{\pi}{4} \leq (m + 1) \arctan \frac{1}{k}$$

を満たす整数 $m$ を探します。これは加法定理を用いて

$$m \arctan \frac{1}{k} = \arctan k_m$$

と表した時、$k_m \leq 1 \leq k_{m+1}$ により判別できます。最後に、加法定理により

$$\begin{align}
m \arctan \frac{1}{k} -\arctan 1 &= \arctan \frac{a_1}{b_1} \\
(m+1) \arctan \frac{1}{k} -\arctan 1 &= \arctan \frac{a_2}{b_2} \\
\end{align}$$

を計算して、$a_1 /b_1$, $a_2 /b_2$ のうち、小さい方を採用します。

In [2]:
from typing import Tuple
import math
from fractions import Fraction

def atan_angle_add(a: Fraction, b: Fraction) -> Fraction:
    return (a + b) /(1 - a * b)

def coefficient(k: int) -> Tuple[int, Fraction, Fraction] :
    m = 0
    prev = 0
    curr = Fraction(1, k)
    
    while curr < 1:
        tmp = curr
        curr = atan_angle_add(curr, Fraction(1, k))
        prev = tmp
        m = m + 1
            
    return m, prev, curr

In [4]:
import sympy

def fraction_prime_factor(f: Fraction) -> str:
    def prime_factor_str(a: int) -> str:
        return " ".join([
            f"{k}**{v}"
            for k, v in sympy.factorint(a).items()
        ])

    return f"{prime_factor_str(f.numerator)} / {prime_factor_str(f.denominator)}"
        
for k in range(2, 14):
    (m, prev, curr) = coefficient(k)
    sub_1 = atan_angle_add(curr, -1)
    sub_2 = atan_angle_add(1, -prev)
    
    print(f"== k = {k} ==")
    
    print(f"{m} arctan 1/{k} + arctan {sub_2}" + (" : ○" if sub_2 < sub_1 else ""))
    print(f"{m+1} arctan 1/{k} - arctan {sub_1}" + (" : ○" if sub_1 < sub_2 else ""))
    
    print("")
    
    print(f"({sub_2})**2 + 1 = {sub_2**2 + 1} = {fraction_prime_factor(sub_2**2 + 1)}")
    print(f"({sub_1})**2 + 1 = {sub_1**2 + 1} = {fraction_prime_factor(sub_1**2 + 1)}")
        
    print("")


== k = 2 ==
1 arctan 1/2 + arctan 1/3
2 arctan 1/2 - arctan 1/7 : ○

(1/3)**2 + 1 = 10/9 = 2**1 5**1 / 3**2
(1/7)**2 + 1 = 50/49 = 2**1 5**2 / 7**2

== k = 3 ==
2 arctan 1/3 + arctan 1/7 : ○
3 arctan 1/3 - arctan 2/11

(1/7)**2 + 1 = 50/49 = 2**1 5**2 / 7**2
(2/11)**2 + 1 = 125/121 = 5**3 / 11**2

== k = 4 ==
3 arctan 1/4 + arctan 5/99 : ○
4 arctan 1/4 - arctan 79/401

(5/99)**2 + 1 = 9826/9801 = 2**1 17**3 / 3**4 11**2
(79/401)**2 + 1 = 167042/160801 = 2**1 17**4 / 401**2

== k = 5 ==
3 arctan 1/5 + arctan 9/46
4 arctan 1/5 - arctan 1/239 : ○

(9/46)**2 + 1 = 2197/2116 = 13**3 / 2**2 23**2
(1/239)**2 + 1 = 57122/57121 = 2**1 13**4 / 239**2

== k = 6 ==
4 arctan 1/6 + arctan 241/1921
5 arctan 1/6 - arctan 475/11767 : ○

(241/1921)**2 + 1 = 3748322/3690241 = 2**1 37**4 / 17**2 113**2
(475/11767)**2 + 1 = 138687914/138462289 = 2**1 37**5 / 7**2 41**4

== k = 7 ==
5 arctan 1/7 + arctan 237/3116
6 arctan 1/7 - arctan 1457/22049 : ○

(237/3116)**2 + 1 = 9765625/9709456 = 5**10 / 2**4 19**2 

$\arctan$ の (加法定理から導かれる) 公式

$$\frac{1}{2} \arctan b = \arctan \frac{\sqrt{b^2 + 1} -1}{b}$$

を用いると、$b^2 -1$ が有理数の平方数ならばより簡単な形に書き換えることができます。例えば $k = 7$ のとき

$$\frac{\pi}{4} = 5\arctan \frac{1}{7} + \arctan \frac{237}{3116}$$

$$\left(\frac{237}{3116}\right)^2 + 1 = \frac{5^{10}}{2^4 \cdot 19^2 \cdot 41^2}$$

なので、

$$ \begin{align}
\arctan \frac{237}{3116} &= 2 \arctan \frac{\left(\frac{5^5}{2^2 \cdot 19 \cdot 41}\right) - 1}{\frac{237}{3116}} \\
&=  2 \arctan \frac{\left(\frac{3125 -3116}{3116}\right)}{\frac{237}{3116}} \\
&=  2 \arctan \frac{9}{237} \\
&=  2 \arctan \frac{3}{79} \\
\end{align}$$

となり、オイラーの公式

$$\frac{\pi}{4} = 5\arctan \frac{1}{7} + 2 \arctan \frac{3}{79}$$

が得られます。

$$\frac{\pi}{4} = 6\arctan \frac{1}{7} -\arctan \frac{1457}{22049}$$

と比べると、

$$\frac{3}{79} = 0.0379... <  0.0660... =\frac{1457}{22049}$$

なので、オイラーの公式の方が収束が速いです。